contracts/cw-proposal-single/src/contract.rs

```diff
@@ -338,6 +369,52 @@ pub fn execute_update_config(
         .add_attribute("sender", info.sender))
 }
 
+pub fn execute_add_hook(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.dao != info.sender {
+        // Only DAO can add hooks
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let validated_address = deps.api.addr_validate(&address)?;
+
+    HOOKS
+        .add_hook(deps.storage, validated_address.clone())
+        .map_err(ContractError::HookError)?;
+
+    Ok(Response::default()
+        .add_attribute("action", "add_hook")
+        .add_attribute("address", validated_address))
```

Comment:
We could lose the clone here by using the initial value. Ignore this if you think it makes the code less readable.

```suggestion
        .add_hook(deps.storage, validated_address)
        .map_err(ContractError::HookError)?;

    Ok(Response::default()
        .add_attribute("action", "add_hook")
        .add_attribute("address", address))
```